### PR TITLE
Support partition commands.

### PR DESCRIPTION
**Pull request by ieure**
This commit adds support for MPD’s partition feature to mpc.

Tested locally and it seems to work fine.  But, it's been over 20 years since I wrote C regularly, so it's exceedingly possible that I've done something ignorant in here.  I'm happy to correct whatever gets pointed out.

### CODE REVIEW

File: src/command.c [hunk 1]
```diff
@@ -1170,3 +1170,57 @@ cmd_replaygain(int argc, char **argv, struct mpd_connection *connection)
 
 	return 0;
 }
+
+int
+cmd_partitionlist(int argc, char **argv, struct mpd_connection *conn)
+{
+    if (!mpd_send_listpartitions(conn)) {
```
Comment (MaxKellermann) [inline src/command.c:1177]: Please use tabs for indentation, like the rest of the code base.

File: src/command.c [hunk 2]
```diff
@@ -1170,3 +1170,57 @@ cmd_replaygain(int argc, char **argv, struct mpd_connection *connection)
...
+    for (int i = 0; i < argc; ++i) {
+        if (!mpd_send_newpartition(conn, argv[i])) {
+            printErrorAndExit(conn);
+        }
+    }
+
+    if (!mpd_command_list_end(conn)) {
+		printErrorAndExit(conn);
+    }
+    return 0;
```
Comment (MaxKellermann) [inline src/command.c:1206]: Here, the `mpd_response_finish()` comand is missing, therefore mpc never evaluates whether this has succeeded.

File: doc/index.rst [hunk 3]
```diff
@@ -371,6 +376,19 @@ Output Commands
...
 
+:command:`moveoutput <output # or name>` - Moves the given output 
+   to the current partition.
+
+Partition Commands
+^^^^^^^^^^^^^^^^^^
+
+:command:`partitions` - Lists all available partitions
+
+:command:`makepart <name>...` - Creates new partitions
```
Comment (MaxKellermann) [inline doc/index.rst:387]: Is it really useful that mpc can create multiple partitions with one command? Usually, a new partition needs to be set up (by moving outputs), which requires additional mpc calls anyway; so not much is saved here.
If you want to reduce overhead, you better have a combined command which creates a partition and sets it up.

File: src/output.c [hunk 4]
```diff
@@ -95,11 +95,12 @@ cmd_outputs(gcc_unused int argc, gcc_unused char **argv,
 		/* We increment by 1 to make it natural to the user  */
 		int id = mpd_output_get_id(output) + 1;
 		const char *name = mpd_output_get_name(output);
+        const char *plugin = mpd_output_get_plugin(output);
 
 		if (mpd_output_get_enabled(output)) {
-			printf("Output %i (%s) is enabled\n", id, name);
+			printf("Output %i (%s: %s) is enabled\n", id, plugin, name);
```
Comment (MaxKellermann) [inline src/output.c:101]: This is unrelated to partitions, and doesn't belong in this commit.

File: src/output.c [hunk 5]
```diff
@@ -279,3 +280,36 @@ cmd_outputset(gcc_unused int argc, char **argv, struct mpd_connection *conn)
...
+        struct mpd_output *output;
+        while ((output = mpd_recv_output(conn)) != NULL) {
+            if (mpd_output_get_id(output) == (unsigned)index) {
+                name = mpd_output_get_name(output);
+                my_finishCommand(conn);
+                break;
+            }
+
+            mpd_output_free(output);
+        }
```
Comment (MaxKellermann) [inline src/output.c:307]: Please move this to a function, e.g. `output_id_to_name()`; this smells like it could be reused some day.

**Review by MaxKellermann — CHANGES_REQUESTED** [review]

File: src/command.c (hunk 1 above)
Comment (ieure) [reply src/command.c:1177]: Fixed.

File: src/command.c (hunk 2 above)
Comment (ieure) [reply src/command.c:1206]: Fixed.

File: src/output.c (hunk 4 above)
Comment (ieure) [reply src/output.c:101]: Fixed.

File: src/output.c (hunk 5 above)
Comment (ieure) [reply src/output.c:307]: I did consider this, but ended up inlining it because `moveoutput` is the only command in the MPD protocol that uses the output name instead of an ID, so I thought it unlikely that it'd be needed elsewhere.

I'll move it if you feel strongly that it should be separate, or there's some output command I missed that needs index->name lookup.

File: doc/index.rst (hunk 3 above)
Comment (ieure) [reply doc/index.rst:387]: I don't feel strongly, but I like how `mpc add` lets me specify multiple things and thought it'd be nice here as well.

Doing the setup in one command seems like it'd require more complex argument parsing than mpc already has, and I didn't feel like hacking on that.  So I think either leave as-is, or change it to only allow creating/deleting one at a time.  Let me know which you'd like.

**Comment by ieure** [comment]
Pinging on this.  Are the changes acceptable now, or do you want further changes?

File: src/main.c [hunk 6]
```diff
@@ -297,7 +297,7 @@ run(const struct command *command, int argc, char **array)
 	if (mpd_connection_cmp_server_version(conn, 0, 21, 0) < 0)
 		fprintf(stderr, "warning: MPD 0.21 required\n");
 
-	if (options.partition != NULL) {
+	if (options.partition != NULL && command->handler != cmd_moveoutput) {
```
Comment (MaxKellermann) [inline src/main.c:300]: This is a very obscure check (comparing function pointers is always obscure and questionable, but for me an acceptable kludge here) and deserves a code comment.

**Comment by MaxKellermann** [comment]
Sorry for the delay, I've had a very busy time at dayjob.

File: src/main.c (hunk 6 above)
Comment (ieure) [reply src/main.c:304]: Added.

**Comment by MaxKellermann** [comment]
Doesn't compile:
```
../../src/output.c: In function ‘cmd_moveoutput’:
../../src/output.c:300:10: error: assignment discards ‘const’ qualifier from pointer target type [-Werror=discarded-qualifiers]
  300 |     name = mpd_output_get_name(output);
      |          ^
../../src/output.c:284:20: error: unused parameter ‘argc’ [-Werror=unused-parameter]
  284 | cmd_moveoutput(int argc, char **argv, struct mpd_connection *conn)
      |                ~~~~^~~~

```

**Comment by ieure** [comment]
> Doesn't compile:

Fixed.  Those didn't fail the compile for me, they were warnings only.  But I agree that it's undesirable to introduce new warnings.

**Comment by ieure** [comment]
Pinging on this again.  I think this is in good shape, but please let me know if you want other changes.

**Comment by MaxKellermann** [comment]
Sorry, I'm busy on dayjob, my response times are even longer than usual.
Your PR fails to build with old libmpdclient versions (down to 2.16 is supported).
```
../../src/main.c: In function 'run':
../../src/main.c:305:8: error: implicit declaration of function 'mpd_run_switch_partition'; did you mean 'mpd_run_sticker_set'? [-Werror=implicit-function-declaration]
   if (!mpd_run_switch_partition(conn, options.partition)) {
```
When you do fixups in a PR that is not yet merged, please amend existing commits.

**Comment by ieure** [comment]
@MaxKellermann This is rebased and commits combined, I believe it's ready.

File: src/main.c [hunk 7]
```diff
@@ -77,6 +77,7 @@ static const struct command {
 	{"crossfade",        0,  1, 0, cmd_crossfade,        "[<seconds>]", "Set and display crossfade settings"},
 	{"current",          0,  0, 0, cmd_current,          "", "Show the currently playing song"},
 	{"del",              0, -1, 1, cmd_del,              "<position>", "Remove a song from the queue"},
+    {"delpart",          1, -1, 0, cmd_partitiondelete,  "<name> ...", "Delete partition(s)"},
```
Comment (MaxKellermann) [inline src/main.c:78]: Bad indent - please use tab, not spaces.

Comment (ieure) [reply src/main.c]: Sorry, thought I'd caught all those.
